enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,363 @@
+  compression level to save data bandwidth. Another, `maxcompcpuusage`, does the same to limit CPU used on compression.
+Neither variable is currently exposed, and both are omitted from this proposal.  (Other per-session performance tuning
+variables like `maxconnrate`, which set the maximum per-process number of connections per second, and `maxsslrate`,
+which sets the maximum per-process number of SSL sessions per second, are not exposed either.)
+
+## Proposal
+
+Currently, OpenShift router's `haproxy-config.template` file contains the lines below in the `defaults` section, but the mechanism for setting the
+environment variables is not exposed.
+```go
+defaults
+...
+{{- if isTrue (env "ROUTER_ENABLE_COMPRESSION") }}
+  compression algo gzip
+  compression type {{ env "ROUTER_COMPRESSION_MIME" "text/html text/plain text/css" }}
+  {{- end }}
+```
+
+### Global HTTP Compression via IngressControllerSpec and environment variables
+
+The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
+```go
+type IngressControllerSpec struct {
+...
+    // httpCompression defines policy for HTTP traffic compression
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy turns on compression for the specified MIME types
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the specified
+// MIME types
+type HTTPCompressionPolicy struct {
+	// compressionMIMETypes is a list of MIME types that should have compression applied.
+	// At least one MIME type must be specified.
+	//
+	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
+	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
+	// formats benefit from compression, but formats that are already compressed (pdf, image,
+	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
+	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=0
+	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`
+}
+
+// CompressionMIMEType defines the format of a single MIME type
+// E.g. "text/css;charset=utf-8", "text/html", "text/*", "image/svg+xml",
+// "application/octet-stream", "X-custom/customsub", etc.
+// The format should follow the Content-Type definition in RFC1341 - https://datatracker.ietf.org/doc/html/rfc1341#page-7
+//
+// +kubebuilder:validation:Pattern="^"?(X\-[^ \(\)<>@,;:\\"\/\[\]\?.\=]+|[application|audio|image|message|multipart|text|video])"?\/"?([^ \(\)<>@,;:\\"\/\[\]\?.\=]+(;? *([^ \(\)<>@,;:\\"\/\[\]\?.\=]+=([^ \(\)<>@,;:\\"\/\[\]\?.\=]|".+")+))?)+"?$
```

Comment:
The quoting with double-quotes looks strange, you don't need to escape things inside brackets, you do need to put any closing bracket first inside brackets (e.g., `[^][ ()<>@,;:"/?.=]`), and you probably want parentheses instead of brackets here: `[application|audio|image|message|multipart|text|video]`.  Do we need to allow double-quotes in the provided MIME type?  Overall, trying to do this validation with a regexp might be more effort than it's worth; it might be better to make things looser, even if that would allow some things that the RFC doesn't strictly allow (such as arbitrary types), but prohibit redundant quoting (if I understand correctly, the yaml syntax might have double-quotes, but the value that is expressed in yaml shouldn't, and the operator should add double-quotes in the generated HAProxy configuration if they are needed there).

Reply:
I'm going to go with this pattern: `^"?(X-[^ ()[\]<>@,;:"/?.=]+|(application|audio|image|message|multipart|text|video))"?/"?(.+)+"?$`